client/clientutil/modelinfo.go

```diff
@@ -0,0 +1,452 @@
+	// it also contains both serial and model assertion headers, but we
+	// follow the same code path for both assertion types and some of the
+	// headers are shared between the two, so it still works out correctly
+	niceOrdering = [...]string{
+		"architecture",
+		"base",
+		"classic",
+		"display-name",
+		"gadget",
+		"kernel",
+		"revision",
+		"store",
+		"system-user-authority",
+		"timestamp",
+		"required-snaps", // for uc16 and uc18 models
+		"snaps",          // for uc20 models
+		"device-key-sha3-384",
+		"device-key",
+	}
+)
+
+type ModelFormatter interface {
+	LongPublisher(storeAccountID string) string
+	GetEscapedDash() string
+}
+
+type PrintModelAssertionOptions struct {
+	TermWidth int
+	AbsTime   bool
+	Verbose   bool
+	Assertion bool
+}
+
+func fmtTime(t time.Time, abs bool) string {
+	if abs {
+		return t.Format(time.RFC3339)
+	}
+	return timeutil.Human(t)
+}
+
+func formatInvalidTypeErr(headers ...string) error {
+	return fmt.Errorf("invalid type for %q header", strings.Join(headers, "/"))
+}
+
+func printVerboseSnapsList(w *tabwriter.Writer, snaps []interface{}) error {
+	printModes := func(snapName string, members map[string]interface{}) error {
+		modes, ok := members["modes"]
+		if !ok {
+			return nil
+		}
+
+		modesSlice, ok := modes.([]interface{})
+		if !ok {
+			return formatInvalidTypeErr("snaps", snapName, "modes")
+		}
+
+		if len(modesSlice) == 0 {
+			return nil
+		}
+
+		modeStrSlice := make([]string, 0, len(modesSlice))
+		for _, mode := range modesSlice {
+			modeStr, ok := mode.(string)
+			if !ok {
+				return formatInvalidTypeErr("snaps", snapName, "modes")
+			}
+			modeStrSlice = append(modeStrSlice, modeStr)
+		}
+		modesSliceYamlStr := "[" + strings.Join(modeStrSlice, ", ") + "]"
+		fmt.Fprintf(w, "    modes:\t%s\n", modesSliceYamlStr)
+		return nil
+	}
+
+	for _, sn := range snaps {
+		snMap, ok := sn.(map[string]interface{})
+		if !ok {
+			return formatInvalidTypeErr("snaps")
+		}
+
+		// Print all the desired keys in the map in a stable, visually
+		// appealing ordering
+		// first do snap name, which will always be present since we
+		// parsed a valid assertion
+		name := snMap["name"].(string)
+		fmt.Fprintf(w, "  - name:\t%s\n", name)
+
+		// the rest of these may be absent, but they are all still
+		// simple strings
+		for _, snKey := range []string{"id", "type", "default-channel", "presence"} {
+			snValue, ok := snMap[snKey]
+			if !ok {
+				continue
+			}
+			snStrValue, ok := snValue.(string)
+			if !ok {
+				return formatInvalidTypeErr("snaps", snKey)
+			}
+			if snStrValue != "" {
+				fmt.Fprintf(w, "    %s:\t%s\n", snKey, snStrValue)
+			}
+		}
+
+		// finally handle "modes" which is a list
+		if err := printModes(name, snMap); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func printVerboseAssertionHeaders(w *tabwriter.Writer, options PrintModelAssertionOptions, assertion asserts.Assertion) error {
```

Comment:
s/Assertion/ModelAssertion/  I know it's used also for serial but that's ok

Reply:
Fixed!